ssas/systems.go

```diff
@@ -74,31 +76,44 @@ type Secret struct {
 
 type ClientToken struct {
 	gorm.Model
-	Label    string `json:"label"`
-	Uuid     string `json:"uuid"`
-	System   System `gorm:"foreignkey:SystemID;association_foreignkey:ID"`
-	SystemID uint   `json:"system_id"`
+	Label     string    `json:"label"`
+	Uuid      string    `json:"uuid"`
+	System    System    `gorm:"foreignkey:SystemID;association_foreignkey:ID"`
+	SystemID  uint      `json:"system_id"`
+	ExpiresAt time.Time `json:"expires_at"`
 }
 
 /*
 	SaveClientToken should be provided with a token label and toke uuid, which will
 	be saved to the client tokens table and associated with the current system.
 */
-func (system *System) SaveClientToken(uuidS string, label string) error {
+func (system *System) SaveClientToken(label string, xData string) (string, error) {
 	db := GetGORMDbConnection()
 	defer Close(db)
 
+	rk, err := NewRootKey(db, CredentialExpiration)
+	if err != nil {
+		return "", fmt.Errorf("could not create a root key for macaroon generation for clientID %s: %s", system.ClientID, err.Error())
+	}
+
+	caveats := make([]Caveats, 3)
+	caveats[0] = map[string]string{"expiration": rk.ExpiresAt.Format(time.RFC3339)}
+	caveats[1] = map[string]string{"system_id": strconv.FormatUint(uint64(system.ID), 10)}
+	caveats[2] = map[string]string{"x_data": base64.StdEncoding.EncodeToString([]byte(xData))}
+
+	token, _ := rk.Generate(caveats, Location)
 	ct := ClientToken{
-		Label:    label,
-		Uuid:     uuidS,
-		SystemID: system.ID,
+		Label:     label,
+		Uuid:      rk.Uuid,
+		SystemID:  system.ID,
+		ExpiresAt: rk.ExpiresAt,
 	}
 
 	if err := db.Create(&ct).Error; err != nil {
-		return fmt.Errorf("could not save client token for clientID %s: %s", system.ClientID, err.Error())
+		return "", fmt.Errorf("could not save client token for clientID %s: %s", system.ClientID, err.Error())
 	}
 	ClientTokenCreated(Event{Op: "SaveClientToken", TrackingID: uuid.NewRandom().String(), ClientID: system.ClientID})
-	return nil
+	return token, nil
```

Comment:
is `token` just the stringified version of `ct`? I'm curious why you save `ct` but return `token`

Reply:
token is what get's generated by the macaroon library, we don't save the macaroon in the database only the uuid tied to the macaroon and the secret used to generate the macaroon.  This is similar to how dpc auth did it, also you can see the original code never returned ct and was returning `ct := b64.StdEncoding.EncodeToString([]byte("client-token-placeholder"))` further down in this file